src/com/amazon/ion/impl/IonReaderBinaryIncremental.java

```diff
@@ -0,0 +1,1932 @@
+     * https://github.com/amzn/ion-java/issues/126 . Support is added now to avoid risking the appearance of performance
+     * degradation if ImportLocation support were added after initial release of this IonReader implementation.
+     */
+    static class ImportLocation {
+
+        // The name of the shared symbol table.
+        final String name;
+
+        // The index into the shared symbol table.
+        final int sid;
+
+        ImportLocation(String name, int sid) {
+            this.name = name;
+            this.sid = sid;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public int getSid() {
+            return sid;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("ImportLocation::{name: %s, sid: %d}", name, sid);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (!(o instanceof ImportLocation)) {
+                return false;
+            }
+            ImportLocation that = (ImportLocation) o;
+            return this.getName().equals(that.getName()) && this.getSid() == that.getSid();
+        }
+
+        @Override
+        public int hashCode() {
+            int result = 17;
+            result += 31 * getName().hashCode();
+            result += 31 * getSid();
+            return result;
+        }
+    }
+
+    /**
+     * SymbolToken implementation that includes ImportLocation.
+     */
+    static class SymbolTokenImpl implements _Private_SymbolToken {
+
+        // The symbol's text, or null if the text is unknown.
+        private final String text;
+
+        // The local symbol ID of this symbol within a particular local symbol table.
+        private final int sid;
+
+        // The import location of the symbol (only relevant if the text is unknown).
+        private final ImportLocation importLocation;
+
+        SymbolTokenImpl(String text, int sid, ImportLocation importLocation) {
+            this.text = text;
+            this.sid = sid;
+            this.importLocation = importLocation;
+        }
+
+        @Override
+        public String getText() {
+            return text;
+        }
+
+        @Override
+        public String assumeText() {
+            if (text == null) {
+                throw new UnknownSymbolException(sid);
+            }
+            return text;
+        }
+
+        @Override
+        public int getSid() {
+            return sid;
+        }
+
+        // Will be @Override once added to the SymbolToken interface.
+        public ImportLocation getImportLocation() {
+            return importLocation;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("SymbolToken::{text: %s, sid: %d, importLocation: %s}", text, sid, importLocation);
+        }
+        @Override
+        public boolean equals(Object o) {
```

Comment:
Correct. That logic doesn't exist at the moment. Because there are two implementations of `SymbolToken` now, but the `SymbolToken` interface doesn't expose `ImportLocation` (there is no `ImportLocation` interface yet), we're not actually using `ImportLocation` in the equality calculation yet. I'll add a comment about this. The Javadoc atop `ImportLocation` explains why it exists even though it isn't exposed or used in equality: since we need it eventually in order to properly preserve correctness on roundtrip, I wanted to add the internal logic that creates import locations now to avoid the appearance of a performance regression if they were added later.

Reply:
Understood. Do we have an issue tracking the eventual work?

Reply:
Yes: #126 .